services/src/main/java/org/fao/geonet/api/records/attachments/AttachmentsApi.java

```diff
@@ -275,22 +276,46 @@ public void getResource(
 
             ApiUtils.canViewRecord(metadataUuid, request);
 
-            response.setHeader("Content-Disposition", "inline; filename=\"" + file.getMetadata().getFilename() + "\"");
-            response.setHeader("Cache-Control", "no-cache");
+            String originalFilename = file.getMetadata().getFilename();
             String contentType = getFileContentType(file.getPath());
-            response.setHeader("Content-Type", contentType);
+            String dispositionFilename = originalFilename;
 
+            // If the image is being resized, always return as PNG and update 
+            // filename and content-type accordingly
             if (contentType.startsWith("image/") && size != null) {
                 if (size >= MIN_IMAGE_SIZE && size <= MAX_IMAGE_SIZE) {
+                    // Set content type to PNG
+                    contentType = "image/png";
+                    // Change file extension to .png
+                    int dotIdx = originalFilename.lastIndexOf('.');
+                    if (dotIdx > 0) {
+                        dispositionFilename = originalFilename.substring(0, dotIdx) + ".png";
+                    } else {
+                        dispositionFilename = originalFilename + ".png";
+                    }
+                    response.setHeader("Content-Disposition", "inline; filename=\"" + dispositionFilename + "\"");
+                    response.setHeader("Cache-Control", "no-cache");
+                    response.setHeader("Content-Type", contentType);
+
+                    // Read, resize, and write the image as PNG, and set Content-Length
                     BufferedImage image = ImageIO.read(file.getPath().toFile());
                     BufferedImage resized = ImageUtil.resize(image, size);
-                    ImageIO.write(resized, "png", response.getOutputStream());
+                    // Write to a byte array first to get the length
+                    ByteArrayOutputStream baos = new ByteArrayOutputStream();
```

Comment:
Have you considered the following 

```
                   CountingOutputStream countingOut = new CountingOutputStream(response.getOutputStream());
                    ImageIO.write(resized, "png", countingOut);
                    response.setContentLengthLong(countingOut.getByteCount());
```

This way you don't have 2 byte arrays in memory which could cause issues if images are larges.

Reply:
The issue is that `response.setContentLengthLong()` must be called before any data is actually written to the HTTP response stream (or at least before the HTTP headers are committed). By the time `countingOut.getByteCount()` gives you the length, it's often too late to set the `Content-Length` header correctly, as the headers might have already been sent.

Reply:
Good point - I will resolve this issue.